README.md
stoarray
========

Library for making api calls to storage arrays with Ruby.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'stoarray'
```

And then execute:

    $ gem install stoarray

## Usage

In your script:

```ruby
require 'stoarray'
```

EMC's Xtremio and Pure's storage arrays are currently supported.

Both api's use json for parameters and the examples below follow suit.
I prefer to set variables that will not change in a json configuration file.
It is very easy to then build from there.

## Clone refresh using Pure
---------------------------

###First the json configuration file:

    {
      "authorization": "123li123o90yourapitoken2h1hi3",
      "base_url": "https://purearray01.something.net/api/1.4/",
      "headers": { "Content-Type": "application/json" },
      "params_snap_u23": {
        "snap_pairs": {
          "x0319t186_u23_1_src": "x0319t186_u23_1_des",
          "x0319t186_u23_2_src": "x0319t186_u23_2_des",
          "x0319t186_u23_3_src": "x0319t186_u23_3_des",
          "x0319t186_u23_4_src": "x0319t186_u23_4_des",
          "x0319t186_u23_5_src": "x0319t186_u23_5_des"
        },
        "source": [
          "x0319t186u23"
        ]
      }
    }

The top three likely will not change between api calls.

+ authorization - This is your api token.
+ base_url      - URL for your array and api version
+ headers       - Pass the content type (JSON in this case)

These are what you need to set for clone refreshes. In this example,
they are nested under "params_snap_u23".

+ snap_pairs    - Map source volume(s) to target volume(s).
+ source        - This is the name of the source protection group.

###Back to your script, after the require 'stoarray'

```ruby
# Location of json configuration file and api token
conf    = JSON.parse(File.read('/Some/path/pure.json'))
token   = conf['authorization']
```

Pure uses cookies. You trade one for your api token and then you can use the cookie
while your session persists (30 minute inactivity timeout, unless you destroy it early).

```ruby
# Get a cookie for our session - required by Pure.
url     = conf['base_url'] + 'auth/session'
headers = conf['headers']
params  = { api_token: token }
cookies = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).cookie

# After we get a cookie, update headers to include it
headers['Cookie'] = cookies
```

Now we will send application type json and the cookie with each call.

```ruby
# Now refresh the clones
params = conf['params_snap_u23']
refresh = Stoarray.new(headers: headers, meth: 'Post', params: params, url: conf['base_url']).refresh
puts "Status:   " + refresh['status'].to_s
puts "Response: " + refresh['response'].to_s
```

In the above example, the source protection group is first snapped.
Next, each target volume is overwritten with the source snapshot (snap_pairs).
Any error along the way will cause the gem to return all status codes
and all array responses. Success gives a 201, SUCCESS.

## Xtremio clone refresh, json first
    {
      "base_url":"https://xmsserver01/api/json/v2/types/",
      "headers": {
        "Content-Type": "application/json",
        "authorization":"Basic alsdkjfsldakjflkdsjflkasdj=="
      },
      "params_refresh_u04": {
        "cluster-id": "xtrmcluster01",
        "from-consistency-group-id": "x0319t186_u04_src",
        "to-snapshot-set-id": "x0319t186_u04_des"
      }
    }

+ base_url        - URL for your array and api
+ headers         - Content type and authorization
+ authorization - "Basic 'Base64 hash of your username and password'"

###Now for your script - this one does a clone refresh

```ruby
#!/usr/bin/env ruby

require 'stoarray'

# Location of json configuration file
conf    = JSON.parse(File.read('/path/to/config/file/xtremioclone.json'))

headers = conf['headers']

# Refresh the snap set
url = conf['base_url'] + 'snapshots'
params  = conf['params_refresh_u04']
refresh = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).refresh
puts "Status:   " + refresh['status'].to_s
puts "Response: " + refresh['response'].to_s
```

## Troubleshooting

error 307 - Enter fully qualified domain name (FQDN) for the array to fix.
    purearray01.something.net instead of just purearray01

Solaris 10 - You may have to specify the path to make/gmake to install the gem.

Error:

```bash
# gem install stoarray 
Fetching: unf_ext-0.0.7.4.gem (100%)
Building native extensions.  This could take a while...
ERROR:  Error installing stoarray:
        ERROR: Failed to build gem native extension.

...followed by...

make failed, exit code 1
```

Fix:

```bash
# which gmake
/usr/sfw/bin/gmake
# export MAKE=/usr/sfw/bin/gmake
# gem install stoarray -s http://rubygems.nordstrom.net
Building native extensions.  This could take a while...

...

it should complete successfully now.
```

A note about the flippy method in array_calls.rb:

    # This method is to get around a "feature" of Xtremio where it renames the
    # target snapshot set during a refresh making it more difficult to automate
    # refreshes as the target keeps changing names. Instead of doing that we
    # flip back and forth between to-snapshot-set-id and to-snapshot-set-id_347.
    # Pass the consistent name, "to-snapshot-set-id" and it flips for you.

Essentially, by design, the array wants to rename the target snapshot set when you refresh. This is not cool when you want to use a consistent name for the set, ie. for automation. The flippy method takes the to-snapshot-set-id from your json or passed as a parameter and then flips back and forth between it and to-snapshot-set-id_347. This way you can keep the same name in your json/scripts, etc.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install` or follow the instructions at the top of the readme.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/kodywilson/stoarray. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

[![Build Status](https://travis-ci.org/kodywilson/stoarray.svg?branch=master)](https://travis-ci.org/kodywilson/stoarray)

[![Gem Version](https://badge.fury.io/rb/stoarray.svg)](https://badge.fury.io/rb/stoarray)

[![Coverage Status](https://coveralls.io/repos/kodywilson/stoarray/badge.svg?branch=master&service=github)](https://coveralls.io/github/kodywilson/stoarray?branch=master)


CHANGELOG.md
# CHANGELOG for stoarray

This file is used to list changes made to the stoarray gem.

## 01022018

* Update README with actual error and fix for Solaris.

## 03182016

* Trying to give a better explanation for the flippy method.
* Update README with flippy explanation and Solaris 10 note.

## 03172016

* All code should have test coverage now!
* Conversion to rest-client complete. Removed getty, del33t, verbal_gerbil.
* make_call - (previously cally) Added rescue so I can return error messages.
* cookie - Removed "testy" stuff & added break on failure (raise).
* Consistent success messages for both Pure and Xtremio clone refreshes.

## 03152016

* Add Mark Gibbon's line to generate an error if a valid cookie is not received.
* All code should be covered now except the cookie block!

## 03142016

* Additional tests using mockable.io: Refresh and flippy covered now.
* New delete method using rest-client. Needs work! (handling params).

## 03112016

* Using mockable.io for API call testing, updating rspec tests.

## 03102016

* Added new array method for gathering stats from the arrays.

## 03092016

* Break call setup out of initialize and into new methods.
* Update dependent methods to use new call methods.
* Add rest-client gem to resolve issue with GETs on Xtremio.

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'coveralls'
Coveralls.wear!

require 'pry'
require 'stoarray'


spec/stoarray_spec.rb
require 'spec_helper'

xtrm_url = 'https://demo7139290.mockable.io/api/json/v2/types/'
pure_url = 'https://demo7139290.mockable.io/api/1.4/'
headers  = { "Content-Type" => "application/json" }
params   = {}

describe Stoarray do

  it "has a version number" do
    expect(Stoarray::VERSION).not_to be nil
  end

  describe ".array" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = pure_url + 'uhrray'
        arrr = Stoarray.new(headers: headers, meth: 'Get', params: params, url: url).array
        expect(arrr['status']).to eql(400)
      end
    end
  end

  describe ".array" do
    context "when asked for the array name" do
      it "returns the array name" do
        url = pure_url + 'array'
        array_name = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).array
        expect(array_name['status']).to eql(200)
      end
    end
  end

  describe ".cookie" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = pure_url + 'auth/sessoon'
        cooky_url = Stoarray.new(headers: headers, meth: 'Get', params: params, url: url).cookie
        expect(cooky_url['status']).to eql(400)
      end
    end
  end

  describe ".cookie" do
    context "when asking for a cookie" do
      it "returns nil if no cookie is returned" do
        url = 'http://www.mocky.io/v2/56ead94d1300001c01737801' + '/auth/session'
        cooky_tst = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).cookie
        expect(cooky_tst).to eql(nil)
      end
    end
  end

  describe ".error_text" do
    context "given three parameters" do
      it "returns a failure code" do
        codey = Stoarray.new.error_text("snap", 'https://wrongo/bongo', "snapshots")
        expect(codey['status']).to eql(400)
      end
    end
  end

  describe ".flippy" do
    context "given a hash containing to-snapshot-set-id" do
      it "returns a flipped snapshot set name" do
        temp_hash = Hash.new
        temp_hash['to-snapshot-set-id'] = "tstserver01"
        url = xtrm_url + 'snapshot-sets'
        flippy = Stoarray.new(headers: headers, url: url).flippy(temp_hash)
        expect(flippy['snapshot-set-name']).to eql('tstserver01_347')
      end
    end
  end

  describe ".flippy" do
    context "given a hash containing to-snapshot-set-id" do
      it "returns a flipped snapshot set name" do
        temp_hash = Hash.new
        temp_hash['to-snapshot-set-id'] = "tstserver02"
        url = xtrm_url + 'snapshot-sets'
        flippy = Stoarray.new(headers: headers, url: url).flippy(temp_hash)
        expect(flippy['to-snapshot-set-id']).to eql('tstserver02_347')
      end
    end
  end

  describe ".host" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = pure_url + 'hast'
        hosty = Stoarray.new(headers: headers, meth: 'Get', params: params, url: url).host
        expect(hosty['status']).to eql(400)
      end
    end
  end

  describe ".host" do
    context "when asked to change a host name" do
      it "returns a 200 on success" do
        url = pure_url + 'host/purehost01'
        hosty = Stoarray.new(headers: headers, meth: 'Put', params: { "name" => "purehost02" }, url: url).host
        expect(hosty['status']).to eql(200)
      end
    end
  end

  describe ".pgroup" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = pure_url + 'pgruop'
        pdaddy = Stoarray.new(headers: headers, meth: 'Get', params: params, url: url).pgroup
        expect(pdaddy['status']).to eql(400)
      end
    end
  end

  describe ".pgroup" do
    context "when asked to snapshot a pgroup" do
      it "returns a 201 on success" do
        params = { :snap => true, :source => 'purevolprd', :suffix => 'interim_snap' }
        url = pure_url + 'pgroup'
        pdaddy = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).pgroup
        expect(pdaddy['status']).to eql(201)
      end
    end
  end

  describe ".refresh" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = xtrm_url + 'sillyputty'
        refreshy = Stoarray.new(headers: headers, meth: 'Get', params: params, url: url).refresh
        expect(refreshy['status']).to eql(400)
      end
    end
  end

  describe ".refresh" do
    context "asked to refresh an Xtremio snapshot set" do
      it "returns a 201 on success" do
        url = xtrm_url + 'snapshots'
        params = { "from-consistency-group-id" => "prdserver01",
                   "to-snapshot-set-id" => "tstserver01" }
        refreshy = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).refresh
        expect(refreshy['status']).to eql(201)
      end
    end
  end

  describe ".refresh" do
    context "asked to refresh a Pure clone set" do
      it "returns a 201 on success" do
        url = pure_url
        params = { "snap_pairs" => {
                   "purevol_1_src" => "purevol_1_des"
                  },
                  "source" => ["purevolprd"]
                  }
        refreshy = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).refresh
        expect(refreshy['status']).to eql(201)
      end
    end
  end

  describe ".refresh" do
    context "asked to refresh a Pure clone set and given an invalid volume" do
      it "returns status 404 on failure" do
        url = pure_url
        params = { "snap_pairs" => {
                   "purevol_1_src" => "purevol_2_des"
                  },
                  "source" => ["purevolprd"]
                  }
        refreshy = Stoarray.new(headers: headers, meth: 'Post', params: params, url: url).refresh
        expect(refreshy['status']['clone_purevol_2_des']['status']).to eql(404)
      end
    end
  end

  describe ".snap" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = xtrm_url + 'snapslots'
        snappy = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).snap
        expect(snappy['status']).to eql(400)
      end
    end
  end

  describe ".snap" do
    context "when asked to show snapshot sets" do
      it "returns a 200 on success" do
        url = xtrm_url + 'snapshot-sets'
        snappy = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).snap
        expect(snappy['status']).to eql(200)
      end
    end
  end

  describe ".snap" do
    context "when asked to show snapshot sets" do
      it "displays the snapshot sets on the array" do
        url = xtrm_url + 'snapshot-sets'
        snappy = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).snap
        x = snappy['response']['snapshot-sets'].any? { |y| y['name'].include?('tstserver01') }
        expect(x).to eql(true)
      end
    end
  end

  describe ".volume" do
    context "given the wrong url" do
      it "returns a failure code" do
        url = xtrm_url + 'voluum'
        voly = Stoarray.new(headers: headers, meth: 'Get', params: {}, url: url).volume
        expect(voly['status']).to eql(400)
      end
    end
  end

  describe ".volume" do
    context "when asked to create a new volume" do
      it "returns a 201 on success" do
        url = pure_url + 'volume/newpurevol01'
        voly = Stoarray.new(headers: headers, meth: 'Post', params: { "size" => "5G" }, url: url).volume
        expect(voly['status']).to eql(201)
      end
    end
  end

end


lib/arraycalls.rb
class Stoarray

  VERBS = {
    'delete' => :Delete,
    'get'    => :Get,
    'post'   => :Post,
    'put'    => :Put
  }

  def initialize(headers: {}, meth: 'Get', params: {}, url: 'https://array/')
    @headers = headers
    @meth    = meth
    @params  = params
    @url     = url
  end

  def array
    if @url.to_s =~ /array/
      responder(make_call)
    else
      error_text("array", @url.to_s, "array")
    end
  end

  def make_call
    @params = @params.to_json unless @meth.downcase == 'get' || @meth.downcase == 'delete'
    begin
      response = RestClient::Request.execute(headers: @headers,
                                             method: VERBS[@meth.downcase],
                                             payload: @params,
                                             timeout: 30,
                                             url: @url,
                                             verify_ssl: false)
    rescue => e
      e.response
    else
      response
    end
  end

  def cookie
    if @url =~ /auth\/session/
      response = make_call
      raise 'There was an issue getting a cookie!' unless response.code == 200
      cookie = (response.cookies.map{|key,val| key + '=' + val})[0]
    else
      error_text("cookie", @url.to_s, 'auth/session')
    end
  end

  def error_text(method_name, url, wanted)
    response = {
      "response" =>
        "ERROR: Wrong url for the #{method_name} method.\n"\
        "Sent: #{url}\n"\
        "Expected: \"#{wanted}\" as part of the url.",
      "status" => 400
    }
  end

  def flippy(temp_hash)
    # This method is to get around a "feature" of Xtremio where it renames the
    # target snapshot set during a refresh making it more difficult to automate
    # refreshes as the target keeps changing names. Instead of doing that we
    # flip back and forth between to-snapshot-set-id and to-snapshot-set-id_347.
    # Pass the consistent name, "to-snapshot-set-id" and it flips for you.
    flippy = temp_hash['to-snapshot-set-id'] + '_347'
    url = 'https://' + URI.parse(@url).host + '/api/json/v2/types/snapshot-sets'
    x = Stoarray.new(headers: @headers, meth: 'Get', params: {}, url: url).snap
    if x['response']['snapshot-sets'].any? { |y| y['name'].include?(flippy) }
      temp_hash['snapshot-set-name']  = temp_hash['to-snapshot-set-id']
      temp_hash['to-snapshot-set-id'] = flippy
    else
      temp_hash['snapshot-set-name']  = flippy
    end
    temp_hash['no-backup'] = true
    temp_hash
  end

  def host
    if @url.to_s =~ /host/
      responder(make_call)
    else
      error_text("host", @url.to_s, "host")
    end
  end

  def pgroup
    if @url.to_s =~ /pgroup/
      responder(make_call)
    else
      error_text("pgroup", @url.to_s, "pgroup")
    end
  end

  def refresh
    case @url.to_s
    when /snapshot/ # Xtremio
      tgt     = @params['to-snapshot-set-id']
      @params = flippy(@params)
      src     = @params['from-consistency-group-id']
      cln     = responder(make_call)
      if cln['status'] == 201
        cln['response'] = 'SUCCESS: Xtremio refresh completed from consistency group ' \
                          + src + ' to snapshot set ' + tgt + '.'
        cln
      else
        cln
      end
    when /1.4/ # Pure, handle the interim snap automagically
      error_state = false
      url     = 'https://' + URI.parse(@url).host + '/api/1.4/pgroup'
      source  = @params['source']
      suffix  = 'interim'
      pam     = { :snap => true, :source => source, :suffix => suffix }
      snap    = Stoarray.new(headers: @headers, meth: 'Post', params: pam, url: url).pgroup
      respond = {
        "response" => {
          "snap" => {
            "response" => snap['response']
          }
        },
        "status" => {
          "snap" => {
            "status" => snap['status']
          }
        }
      }
      if snap['status'] >= 200 && snap['status'] <= 299
        @params['snap_pairs'].each do |key, val|
          tgt = 'clone_' + val # used only for logging!
          src = source[0] + '.' + suffix + '.' + key
          pam = { :overwrite => true, :source => src }
          url = 'https://' + URI.parse(@url).host + '/api/1.4/volume/' + val
          clone = Stoarray.new(headers: @headers, meth: 'Post', params: pam, url: url).volume
          respond['response'][tgt] = {
            "response" => clone['response']
          }
          respond['status'][tgt] = {
            "status" => clone['status']
          }
        end
        url   = 'https://' + URI.parse(@url).host + '/api/1.4/pgroup/' + source[0] + '.' + suffix
        zappy = Stoarray.new(headers: @headers, meth: 'Delete', params: {}, url: url).pgroup
        respond['response']['destroy'] = {
          "response" => zappy['response']
        }
        respond['status']['destroy'] = {
          "status" => zappy['status']
        }
        url = url + '?eradicate=true'
        disintegrate = Stoarray.new(headers: @headers, meth: 'Delete', params: {}, url: url).pgroup
        respond['response']['eradicate'] = {
          "response" => disintegrate['response']
        }
        respond['status']['eradicate'] = {
          "status" => disintegrate['status']
        }
      end
      respond['status'].each do |key, val|
        error_state = true if val.any? { |status, code| code.to_i < 200 || code.to_i > 299 }
      end
      if error_state == true
        respond
      else
        response = {
          "response" =>
            "SUCCESS: Pure refresh completed from #{source[0]} protection group.\n",
          "status" => 201
        }
      end
    else
      error_text("refresh", @url.to_s, "snapshot or 1.4")
    end
  end

  def responder(response)
    response = {
      "response" => JSON.parse(response.body),
      "status" => response.code.to_i
    }
  end

  def snap
    if @url.to_s =~ /snapshot/
      responder(make_call)
    else
      error_text("snap", @url.to_s, "snapshot")
    end
  end

  def volume
    if @url.to_s =~ /volume/
      responder(make_call)
    else
      error_text("volume", @url.to_s, "volume")
    end
  end

end
